modules/criteoBidAdapter.js

```diff
@@ -19,20 +20,241 @@ import { ajax } from '../src/ajax.js';
+
+  imp.tagid = bidRequest.adUnitCode;
+  deepSetValue(imp, 'ext', {
+    ...bidRequest.params.ext,
+    ...imp.ext,
+    rwdd: imp.rwdd,
+    floors: getFloors(bidRequest),
+    bidder: {
+      publishersubid: params?.publisherSubId,
+      zoneid: params?.zoneId,
+      uid: params?.uid,
+    },
+  });
+
+  delete imp.rwdd // oRTB 2.6 field moved to ext
+  delete imp.ext.tid // field moved to source
+
+  if (!context.fledgeEnabled && imp.ext.igs?.ae) {
+    delete imp.ext.igs.ae;
+  }
+
+  if (hasVideoMediaType(bidRequest)) {
+    const paramsVideo = bidRequest.params.video;
+    if (paramsVideo !== undefined) {
+      deepSetValue(imp, 'video', {
+        ...imp.video,
+        skip: imp.video.skip || paramsVideo.skip || 0,
+        placement: imp.video.placement || paramsVideo.placement,
+        minduration: imp.video.minduration || paramsVideo.minduration,
+        playbackmethod: imp.video.playbackmethod || paramsVideo.playbackmethod,
+        startdelay: imp.video.startdelay || paramsVideo.startdelay || 0,
+      })
+    }
+    deepSetValue(imp, 'video.ext', {
+      context: bidRequest.mediaTypes.video.context,
+      playersizes: parseSizes(deepAccess(bidRequest, 'mediaTypes.video.playerSize'), parseSize),
+      plcmt: bidRequest.mediaTypes.video.plcmt,
+      poddur: bidRequest.mediaTypes.video.adPodDurationSec,
+      rqddurs: bidRequest.mediaTypes.video.durationRangeSec,
+    })
+  }
+
+  if (imp.native && typeof imp.native.request !== 'undefined') {
+    let requestNative = JSON.parse(imp.native.request);
+
+    // We remove the native asset requirements if we used the bypass to generate the imp
+    const hasAssetRequirements = requestNative.assets &&
+      (requestNative.assets.length !== 1 || Object.keys(requestNative.assets[0]).length);
+    if (!hasAssetRequirements) {
+      delete requestNative.assets;
+    }
+
+    deepSetValue(imp, 'native.request_native', requestNative);
+    delete imp.native.request;
+  }
+
+  return imp;
+}
+
+/**
+ * Builds a request object for the ORTB 2.5 request.
+ *
+ * @param {function} buildRequest - The function for building a request object.
+ * @param {Array} imps - An array of ORTB 2.5 impression objects.
+ * @param {Object} bidderRequest - The bidder request object.
+ * @param {Object} context - The context object.
+ * @returns {Object} The ORTB 2.5 request object.
+ */
+function request(buildRequest, imps, bidderRequest, context) {
+  let request = buildRequest(imps, bidderRequest, context);
+
+  if (typeof request.app !== 'undefined') {
+    // app
+    if (typeof context.publisherId !== 'undefined') {
+      deepSetValue(request, 'app.publisher.id', context.publisherId);
+    }
+  } else {
+    // site
+    if (typeof context.url !== 'undefined') {
+      deepSetValue(request, 'site.page', context.url);
+    }
+    if (typeof context.publisherId !== 'undefined') {
+      deepSetValue(request, 'site.publisher.id', context.publisherId);
+    }
+  }
+
+  if (bidderRequest) {
```

Comment:
Much of this logic is already included in FPD (`ortb2` / `ortb2Imp`) and/or ortbConverter (which automatically copies everything from FPD). There's a difference in how you handle transaction IDs (more on that below). If you want to compare:

- coppa - https://github.com/prebid/Prebid.js/blob/542694585269e3796d65205283783915a8cf9f70/src/fpd/enrichment.js#L138-L140
- us_privacy - https://github.com/prebid/Prebid.js/blob/542694585269e3796d65205283783915a8cf9f70/modules/consentManagementUsp.js#L249-L255
- gdpr (but not `gdprversion`, only `regs.ext.gdpr` and `user.ext.consent`) - https://github.com/prebid/Prebid.js/blob/542694585269e3796d65205283783915a8cf9f70/modules/consentManagement.js#L307-L314 
- gpp  - https://github.com/prebid/Prebid.js/blob/542694585269e3796d65205283783915a8cf9f70/modules/consentManagementGpp.js#L510-L521
- eids - https://github.com/prebid/Prebid.js/blob/542694585269e3796d65205283783915a8cf9f70/modules/userId/index.js#L1101-L1106
- schain - https://github.com/prebid/Prebid.js/blob/542694585269e3796d65205283783915a8cf9f70/modules/schain.js#L198-L205

And on the response side, dsa: https://github.com/prebid/Prebid.js/blob/542694585269e3796d65205283783915a8cf9f70/libraries/ortbConverter/processors/default.js#L100-L102


From transaction IDs, `source.tid` is an auction ID and transaction (impression) IDs are in `imp[].ext.tid` - see [this summary](https://docs.prebid.org/dev-docs/pb8-notes.html#transaction-identifiers-are-now-reliable-and-opt-in)

If you decide to use the built-in logic you'll probably have to update unit tests - there's [some helper utilities](https://github.com/search?q=repo%3Aprebid%2FPrebid.js%20addFPDToBidderRequest&type=code) to run the FPD enrichment logic (and the test suite should import all the relevant modules).

Reply:
Thank you for the very detailed comparisons, much appreciated!

I pushed a change to use built-in modules as it looks more convenient.
For transaction ids, it should be fixed: we'll forward both `imp[].ext.tid` from `ortb2Imp` and `source.tid` from `ortb2` and decide how to properly deal with them on our backend.

As some modules aren't enabled by default (unlike fpd), there's nothing special to do on the adapter to enable those?

Reply:
> As some modules aren't enabled by default (unlike fpd), there's nothing special to do on the adapter to enable those?

No - if the publisher includes them you get the logic (and if they don't include them, but they still want to "manually" provide info such as for example gpp strings, they can only do it through FPD which is still included).